odl/contrib/solvers/spdhg/misc.py

```diff
@@ -0,0 +1,857 @@
+            - f(v) + r.inner(v))
+
+
+def divide_1Darray_equally(ind, n_subsets):
+    """
+    Divide an array into equal chunks to be used for instance in OSEM.
+
+    Parameters
+    ----------
+    ind : ndarray
+        input array
+    nsubsets : int
+        number of subsets to be divided into
+
+    Returns
+    -------
+    subset2ind : list
+        list of indices for each subset
+    ind2subset : list
+        list of subsets for each index
+    """
+
+    n_ind = len(ind)
+    subset2ind = []
+
+    for i in range(n_subsets):
+        subset2ind.append([])
+        for j in range(i, n_ind, n_subsets):
+            subset2ind[i].append(ind[j])
+
+    ind2subset = []
+    for i in range(n_ind):
+        ind2subset.append([])
+
+    for i in range(n_subsets):
+        for j in subset2ind[i]:
+            ind2subset[j].append(i)
+
+    return (subset2ind, ind2subset)
+
+
+def TV(domain, grad=None):
+    """ Total variation functional.
+
+    Parameters
+    ----------
+    domain : odlspace
+        domain of TV functional
+    grad : gradient operator, optional
+        Gradient operator of the total variation functional. This may be any
+        linear operator and thereby generalizing TV. default=forward
+        differences with Neumann boundary conditions
+
+    Examples
+    --------
+    Check that the total variation of a constant is zero
+
+    >>> import odl.contrib.spdhg as spdhg, odl
+    >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+    >>> TV = spdhg.TV(space)
+    >>> x = space.one()
+    >>> TV(x) < 1e-10
+    """
+
+    if grad is None:
+        grad = odl.Gradient(domain, method='forward', pad_mode='symmetric')
+        grad.norm = 2 * np.sqrt(sum(1 / grad.domain.cell_sides**2))
+    else:
+        grad = grad
+
+    f = odl.solvers.GroupL1Norm(grad.range, exponent=2)
+
+    return f * grad
+
+
+class TV_NonNegative(odl.solvers.Functional):
+    """ Total variation function with nonnegativity constraint and strongly
+    convex relaxation.
+
+    In formulas, this functional may represent
+
+        alpha * |grad x|_1 + char_fun(x) + beta/2 |x|^2_2
+
+    with regularization parameter alpha and strong convexity beta. In addition,
+    the nonnegativity constraint is achieved with the characteristic function
+
+        char_fun(x) = 0 if x >= 0 and infty else.
+
+    Parameters
+    ----------
+    domain : odlspace
+        domain of TV functional
+    alpha : scalar, optional
+        Regularization parameter, positive
+    prox_options : dict, optional
+        name: string, optional
+            name of the method to perform the prox operator, default=FGP
+        warmstart: boolean, optional
+            Do you want a warm start, i.e. start with the dual variable
+            from the last call? default=True
+        niter: int, optional
+            number of iterations per call, default=5
+        p: array, optional
+            initial dual variable, default=zeros
```

Comment:
The good thing about the dictionary is that it is very easy to distinguish which parameters are for what without actually changing their name. Thus, I am not sure which of the two is the better option.

Reply:
Sure, but I find the pain to create a dictionary greater than the win in nice parameter separation. A middle ground may be to put them all in `kwargs` and name them `prox_p`, `prox_niter` etc. Also very clear IMO.

Reply:
I agree that this would be clear, too, but I am still in favour of the dict method. In any case I think this design choice is not very important here. 

More interesting would be to me if odl should support this kind of structure (iterative algorithms compute prox operators) and if so, how.

Reply:
> More interesting would be to me if odl should support this kind of structure (iterative algorithms compute prox operators) and if so, how.

In an earlier version of the solvers, we supplied the prox operator (factory) directly. That was before the functionals came up. There it would be obvious how to do it:

- Wrap the iterative method for the prox into an `Operator` (parametrized by `sigma`)
- Make a function that takes `sigma` and returns the `prox[sigma]` operator.

Now with the current (IMO better) API of supplying functionals to the solvers, the same can be done, just with one more level of indirection: The prox function has to be wrapped into a `Functional` property.

This also means that parameters for the prox iteration need to be given when the functional is created, since no parameters can be provided later on.